src/pn532.cpp

```diff
@@ -0,0 +1,296 @@
+/*
+ * Author: Matthias Akstaller
+ */
+
+#if defined(ENABLE_DEBUG) && !defined(ENABLE_DEBUG_NFCREADER)
+#undef ENABLE_DEBUG
+#endif
+
+#include "pn532.h"
+#include <Wire.h>
+#include "app_config.h"
+#include "debug.h"
+#include "lcd.h"
+
+#ifndef I2C_SDA
+#define I2C_SDA 21
+#endif
+
+#ifndef I2C_SCL
+#define I2C_SCL 22
+#endif
+
+#define  SCAN_FREQ            1000
+
+#define MAXIMUM_UNRESPONSIVE_TIME  60000UL //after this period the pn532 is considered offline
+#define AUTO_REFRESH_CONNECTION         30 //after this number of polls, the connection to the PN532 will be refreshed
+
+
+/*
+ * Documentation of the I2C messages:
+ * NXP PN532 manual, https://www.nxp.com/docs/en/user-guide/141520.pdf
+ * (accessed: 2022-01-14)
+ */
+
+#define PN532_I2C_ADDRESS  (0x48 >> 1)
+#define PN532_RDY 0x01
+#define PN532_FRAME_HEADER_LEN 4
+#define PN532_FRAME_IDENTIFIER_LEN 1
+#define PN532_CMD_SAMCONFIGURATION_RESPONSE 0x15
+#define PN532_CMD_INAUTOPOLL_RESPONSE 0x61
+
+#define I2C_READ_BUFFSIZE 35 //might increase later
+
+PN532::PN532() : MicroTasks::Task() {
+    
+}
+
+void PN532::begin() {
+    Wire.begin(I2C_SDA, I2C_SCL);
+    MicroTask.startTask(this);
+}
+
+unsigned long PN532::loop(MicroTasks::WakeReason reason){
+
+    read();
+
+    if (!config_rfid_enabled()) {
+        status = DeviceStatus::NOT_ACTIVE;
+        return SCAN_FREQ;
+    }
+
+    if (status == DeviceStatus::ACTIVE && millis() - lastResponse > MAXIMUM_UNRESPONSIVE_TIME) {
+        DBUGLN(F("[rfid] connection to PN532 lost"));
+        lcd.display("RFID chip not found", 0, 1, 5 * 1000, LCD_CLEAR_LINE);
+        status = DeviceStatus::FAILED;
+    }
+    
+    if (status != DeviceStatus::ACTIVE) {
+        initialize();
+        return SCAN_FREQ;
+    }
+
+    if (pollCount >= AUTO_REFRESH_CONNECTION) {
+        initialize();
+        pollCount = 0;
+    } else {
+        poll();
+        pollCount++;
+    }
+
+    return SCAN_FREQ;
+}
+
+bool PN532::readerFailure() {
+    return config_rfid_enabled() && status == DeviceStatus::FAILED;
+}
+
+void PN532::initialize() {
+
+    /*
+     * Command hardcoded according to NXP manual, page 89
+     */
+    uint8_t SAMConfiguration [] = {0x00, 0xFF, 0x05, 0xFB, 
+                                   0xD4, 0x14, 0x01, 0x00, 0x00,
+                                   0x17};
+    Wire.beginTransmission(PN532_I2C_ADDRESS);
+    Wire.write(SAMConfiguration, sizeof(SAMConfiguration) / sizeof(*SAMConfiguration));
+    Wire.endTransmission();
+
+    listen = true;
+    //"flush" ACK
+    delay(30);
```

Comment:
Shouldn't use `delay` should set some state and drop out of `loop` with a return of `30` and perform the action next time round, as it happens `read` is the first thing that is done in the loop so might be ok

Reply:
Okay, will change it